### ISSUE TITLE
Separate Node into "Links" and "Data"

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida_core#856]
Writeup of a discussion at the coding week:

In the current implementation, each ``Node`` contains basically two types of information:

* Data (attributes), which describe the node's content and cannot be changed after it was stored.
* Links and other metadata, which describes how the node was created and its relationship in the graph.

The goal of this proposal is to separate these two parts:

* "Data Nodes" contain the Data
* "Reference Nodes" contain the metadata, and a reference to the "Data Node".

This setup would allow multiple Reference Nodes to share the same Data Node, meaning that a "shallow copy" of the Node is possible. This has significant impact on how caching can be done: Currently, a Node needs to be copied fully, because otherwise the "shared" Node would potentially create a loop in the provenance graph. By allowing a "shallow copy" (keeping the Data Node, but creating a new Reference Node), the graph would remain acyclic, while also avoiding a full copy of the (potentially heavy) data.

Obviously, the public interface of the ``Node`` should remain unchanged, and the splitting into reference and data node should be an implementation detail, up to the fact that it allows a "shallow copy" method.
For nodes which are "shallow copied", this should be indicated via a new link type (which should NOT be followed when constructing the provenance graph).

**Comment by giovannipizzi** [comment]
I think the simplest approach is to keep DbNode for the actual links, and move the data (attributes, extras, and reference to the folders) to a DbNodeData or similar. DbNode would have a foreign key to DbNodeData. 
- we need probably to have a 'purge' functionality to clean up DbNodeData that has no references
- probably we need to have two UUIDs, one for DbNode (visible to users) and one for DbNodeData (e.g. used for naming folders in the repository?)
- Where should label, description and type be? for type, think also to efficiency (maybe it should be in the Data, but then there is an additional join to do every time

I also think that a good criterion is: when we will delete nodes, if we want to leave a reference (a "ghost node"), we could just leave the DbNode entry and clear (set to `null`) the foreign key to DbNodeData?

**Comment by greschd** [comment]
On the "purge" functionality, I agree that there should be an option to do this manually, but shouldn't the general way in which data is deleted be by doing reference counting? That is, the DbNode notifies the DbNodeData when it gets constructed / destructed, which triggers automatic destruction when the reference count reaches zero?

**Comment by giovannipizzi** [comment]
Possibly yes. There was a discussion on whether to convert a node to be a 'ghost' without actually deleting the data - there are however no specs for this yet, so we keep this as a note in this issue for future reference.

**Comment by greschd** [comment]
By default I'm not sure the 'ghost nodes' would be a good idea, since I'd guess the total number of DbNodes influences the query performance quite strongly. Maybe it could be optional? But then we could still do reference counting, i.e. if all the DbNodes referring to a specific DbNodeData have 'ghosted' the data can be deleted. Of course while there are other DbNodes keeping it alive we could also keep the DbNodeData available, but essentially turning it into a weak reference.

**Comment by giovannipizzi** [comment]
Yeah, the discussion is a bit more complex than that. In my opinion, we shouldn't just have a lot of ghost nodes - we should only keep ghost nodes that are direct inputs or direct outputs of a calculation. As soon as a ghost node is attached (only) to another ghost node, it can be deleted (i.e. if you delete the calculation itself, you also delete all the connected ghost nodes).

The reasoning is that if you delete an input of a calculation and don't leave a trace of this, looking back into the DB would result into a very 'weird' sealed calculation, where it run but with missing inputs. (also, BTW, we need to think about the effects of deleting inputs for caching).

Another thing to make sure is that if I have "Data1 -> Calc -> Data2" and I delete Calc, I think that Data1 and Data2 can stay (and no ghost nodes are needed) but if later I import back only Calc, it would import with all inputs and outputs as ghosts, so "GhostData1 -> Calc -> GhostData2" but then it should recognise that Data1 and Data2 are there, and replace the Ghost with the actual data, reconnecting Data1 and Data2 correctly and re-establishing provenance.

**Comment by greschd** [comment]
Yeah, I guess that opens quite a can of worms. From a "pure" view it shouldn't even be allowed to delete nodes without also deleting all its descendants, but that might well be too restrictive in practice.

For caching:
- if the hash still exists, we don't really need to have the inputs, we can still look up the calculation
- if it needs to be re-hashed, the only reasonable thing to do is to set the hash to ``None`` / invalid
- not sure what to do about when _outputs_ are deleted -- is this communicated to the calculation itself in some way? If yes, that should invalidate the hash.

**Comment by greschd** [comment]
But I think we're actually discussing three separate issues here:
- the link / data separation of nodes
- (related) if "ghost" link nodes should be allowed, and how to handle them
- what happens to the parents / children when nodes are deleted
